_posts/2024-09-24-creating-a-new-database-integration-with-data-prepper.markdown

```diff
@@ -0,0 +1,393 @@
+Anything that a user would need to provide to run your plugin should be defined in this configuration. 
+
+To get started, see this [sample source example](https://github.com/graytaylor0/data-prepper/blob/SourceCoordinationSampleSource/data-prepper-plugins/sample-source-coordination-source/src/main/java/SampleSource.java). 
+This example is for a hypothetical database, where the only configurations required are `database_name`, `username`, and `password` ([Configuration Class](https://github.com/graytaylor0/data-prepper/blob/SourceCoordinationSampleSource/data-prepper-plugins/sample-source-coordination-source/src/main/java/SampleSourceConfig.java)). Note in the example code that the plugin name and configuration class is defined in the `@DataPrepperPlugin` annotation on the class. 
+When running Data Prepper, the pipeline.yaml for this source would be as follows:
+
+```yaml
+version: 2
+sample-source-pipeline:
+  source:
+   sample_source:
+     database_name: "my-database"
+     username: 'my-username'
+     password: 'my-password'
+  sink:
+    - stdout:
+```
+
+### Using the Source Coordination APIs
+
+The [Source Coordination Interface](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/enhanced/EnhancedSourceCoordinator.java) defines 
+the methods available for interacting with the [Source Coordination Store](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/SourceCoordinationStore.java).
+
+These methods enable plugin creators to handle CRUD operations on their partitions, as well as acquiring the next partition to be worked on with the `acquireAvailablePartition(String partitionType)` method. A common pattern when using Source Coordination
+is to assign a "leader" Data Prepper container that is responsible for partition discovery and creation of partitions. This is achieved by creating one partition on startup of Data Prepper that is the "Leader Partition", and utilizing the `acquireAvailablePartition(LeaderPartition.PARTITION_TYPE)`
+method to determine if the job of partition discovery and creation of other partitions is assigned to a given Data Prepper container. 
+
+The following code snippet demonstrates a simple flow while using source coordination. This example references a hypothetical database, where the "partition" is a single database file.
+The full code for this example can be found [here](https://github.com/graytaylor0/data-prepper/blob/6e38dead8e9beca089381519654f329b82524b9d/data-prepper-plugins/sample-source-coordination-source/src/main/java/DatabaseWorker.java#L40).
+
+The overview of this code is as follows for the different numbered sections
```

Comment:
It's referring to the numbered comments in the code snippet. Ex: `// 1 - Check if this node is already the leader. If it is not, then try to acquire leadership in case the leader node has crashed`